{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "# Metaheuristic Algorithms\n",
    "\n",
    "##  Index\n",
    "1. Problem\n",
    "1. Review and Implementation Algorithms\n",
    "    1. GA\n",
    "    1. DE \n",
    "    1. PSO\n",
    "1. Evaluation\n",
    "    1. Performance\n",
    "        1. Accuracy\n",
    "        1. Time\n",
    "        1. Memory\n",
    "\n",
    "## 1. Probrem\n",
    "Try to find the minimum points of the following functions, using GA, DE, and PSO and compare their performance\n",
    "\n",
    "$$f_1(x_1,x_2,x_3)=\\sum_{i=1}^3 x_i^2, 5.11 \\leq x_i \\leq 5.12$$\n",
    "\n",
    "$$f_2(x_1,x_2)=100(x_1^2 - x_2)^2 + (1-x_1)^2, -2.047 \\leq x_i  \\leq 2.048 $$\n",
    "\n",
    "![3Dgraph](report_image/1.png)\n",
    "![3Dgraph](report_image/2.png)\n",
    "\n",
    "$F1 min : 0.0 \\; (0.01039) \\; \\; \\; F2 min : 0.0 \\;(0.07062$)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "### 2. Review and Implementation algorithms (keypoint)\n",
    "- ####  Genetic Algorithm (GA)\n",
    "Selection, crossover and mutation\n",
    "![FC_GA](report_image/FC_GA.png)\n",
    "> cite: Handout Lec06\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "#### Reconstruction of the phenotype from the genotype site\n",
    "\n",
    ">$y_{i}=\\sigma_{k=1}^{B}b_{ik}2^{-k},    \\;\\; w_{i}=a[y_{i}]+b$\n",
    "\n",
    "> where $a$ and $b$ are the scaling and shifting factors.\n",
    "\n",
    ">cite: Handout Lec06 p.22\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "class Gene:    \n",
    "    #-------------------------------------------------------\n",
    "    # This class has a genotype(M bits binary) and a phenotype (decimal number collesponding the genotype)\n",
    "    #\n",
    "    # obj: objective function\n",
    "    # M : the length of genotype\n",
    "    # dim : the dimension of the solution space \n",
    "    # a : scaling\n",
    "    # b : shifting\n",
    "    #-------------------------------------------------------\n",
    "    \n",
    "    def __init__(self, obj, M,dim,a,b, genotype=None):\n",
    "        if genotype is None:\n",
    "            # generate M-bits binary dim times\n",
    "            self.genotype = [[int(g[k]%2) for k in range(0,M) for g in [random(M)*10]] for x in range(0,dim)]\n",
    "            # decode the genotype\n",
    "            self.phenotype = [sum([g[i]*2**(-i) for i in range(0,len(g))])*a-b for g in self.genotype]\n",
    "            self.fitness = obj(self.phenotype)\n",
    "        else:\n",
    "            self.genotype = genotype\n",
    "            self.phenotype = [sum([g[i]*2**(-i) for i in range(0,len(g))])*a-b for g in self.genotype]\n",
    "            self.fitness = obj(self.phenotype)   \n",
    "    \n",
    "    def getGenotype(self):\n",
    "        return self.genotype\n",
    "\n",
    "    def getPhenotype(self):\n",
    "        return self.phenotype   \n",
    "    \n",
    "    def setFitness(self, phenotype):\n",
    "        self.fitness = obj(self.phenotype)\n",
    "\n",
    "    def setPhenotype(self,genotype):\n",
    "        self.phenotype = [sum([g[i]*2**(-i) for i in range(0,len(g))])*a-b for g in genotype]\n",
    "        self.setFitness(self.phenotype)\n",
    "        \n",
    "    def setGenotype(self,genotype):\n",
    "        self.genotype = genotype\n",
    "        self.setPhenotype(self.genotype)\n",
    "     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def GA(obj,dim) :\n",
    "    N = 10 # the number of population\n",
    "    M = 10 # the lenght of genotype\n",
    "    sr = 0.7 # selection rate\n",
    "    a = 2.5\n",
    "    b = 0\n",
    "    popSize = int(N*sr) # the number of trucation selection (correspond to selection rate)\n",
    "    pm = 0.6\n",
    "    \n",
    "    def crossover(chromosomes):  \n",
    "        crossover_point = randint(0,len(chromosomes[0]))    \n",
    "        temp = chromosomes[0][crossover_point:]\n",
    "        chromosomes[0][crossover_point:] = chromosomes[1][crossover_point:]\n",
    "        chromosomes[1][crossover_point:] = temp\n",
    "        return chromosomes\n",
    "\n",
    "    def mutation(gene,mutation_rate):\n",
    "        dim = len(gene.genotype)\n",
    "        def generate_mutation(genotype,pm):\n",
    "            return [(bit+int(random()>pm))%2 for bit in genotype]    \n",
    "        gene.setGenotype([generate_mutation(gene.genotype[i],mutation_rate) for i in range(0,dim)])\n",
    "    \n",
    "    #Step 1 Initialization\n",
    "    population = [Gene(obj,M,dim,a,b) for g in range(0,N)] # Step 2 Evaluation\n",
    "\n",
    "\n",
    "    for cnt in range(0,300):\n",
    "        # Step 3.1 Selection (by tounament)\n",
    "        new_population = sorted(population, key=lambda population: population.fitness) # sorted by fitness\n",
    "        [new_population.pop(-1) for i in range(0,popSize)] # trunction selection  \n",
    "       \n",
    "        # Step 3.2 Reproduction      \n",
    "        # for each new population\n",
    "        for i in range(N-popSize,len(population)):\n",
    "            # select parents randomly             \n",
    "            parents = [new_population[randint(0,N-popSize)] for i in range(0,2)]\n",
    "            # produce children genotype by crossover the parents\n",
    "            children_geno = [crossover([parent.genotype[i] for i in range(0,dim)]) for parent in parents][randint(0,2)]\n",
    "            new_population.append(Gene(obj,M,dim,a,b,children_geno))     \n",
    "        [mutation(p,pm) for p in population]\n",
    "        population = new_population\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### F1\n",
    "![GA_objhistory_f1](report_image/GA_objhistory_f1.png)\n",
    "![GA_objhistory_f1](report_image/GA_cnt_history_f1.png)\n",
    "#### F2\n",
    "![GA_objhistory_f2](report_image/GA_objhistory_f2.png)\n",
    "![GA_objhistory_f2](report_image/GA_cnt_history_f2.png)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Differential Evolution (DE)\n",
    "Perubtation is caused by 3 agents $r_{1},r_{2},r_{3}$.\n",
    "\n",
    "\n",
    "$u_i = r_{3i} + F(r_{1i} - r_{2_i}), \\;\\; i=1,2,...,D$\n",
    "![FC_GA](report_image/FC_DE.png)\n",
    "> cite: Handout Lec08"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "class Agent:\n",
    "    def __init__(self,obj,vector):\n",
    "        self.vector = vector\n",
    "        self.fitness = obj(vector)\n",
    "    \n",
    "    def getVector(self):\n",
    "        return self.vector\n",
    "\n",
    "    def setVector(self, vector):\n",
    "        self.vector = vector"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def DE(obj,dim) :\n",
    "    N = 10\n",
    "    r = 5.11\n",
    "    w = 1 # differential weight [0,2]\n",
    "    CR = 0.6 # crossover rate [0.4,1]\n",
    "    val_his = []\n",
    "    pos_his = []\n",
    "    \n",
    "    def new_agent(obj,agent, population, weight, crossover_rate):\n",
    "        N = len(population)\n",
    "        dim = len(agent.vector)      \n",
    "    \n",
    "        i_1 = randint(0,N) # FIX: durty code\n",
    "        i_2 = randint(0,N)\n",
    "        i_3 = randint(0,N)\n",
    "    \n",
    "        while ((i_2 == i_3) or (i_1 == i_2) or (population[i_1] == agent)):\n",
    "            i_3 = randint(0,N)\n",
    "            while ((i_1 == i_2) or (population[i_1] == agent)):\n",
    "                i_2 = randint(0,N)\n",
    "                while (population[i_1] == agent):\n",
    "                    i_1 = randint(0,N)\n",
    "    \n",
    "        new_vec = [population[i_3].vector[i] + weight*(population[i_1].vector[i]-population[i_2].vector[i]) for i in range(0,dim)]\n",
    "        new_vec = [new_vec[i] if (random() < crossover_rate) else agent.vector[i]  for i in range(0,dim)]\n",
    "        new_p = Agent(obj,new_vec)\n",
    "        return new_p\n",
    "    \n",
    "    \n",
    "    # Step 1. Initialization\n",
    "    population= [Agent(obj,RV4LimitSpace(dim,r)) for i in range(0,N)]\n",
    "    \n",
    "    # Step 2. Get an agent\n",
    "    # obtain the best fitness agent by sorting and get an head element.\n",
    "    best_agent = sorted(population, key=lambda population: population.fitness)[0]\n",
    "\n",
    "    # update information (not essential part)\n",
    "    val = best_agent.fitness\n",
    "    val_his.append(val)\n",
    "    solution = best_agent.vector\n",
    "    pos_his.append(solution)\n",
    "        \n",
    "    for cnt in range(0,300):\n",
    "        new_population = []\n",
    "\n",
    "        # Step 3. Perturbation\n",
    "        candidates = [new_agent(obj,population[0],population, w,CR) for p in population]\n",
    "   \n",
    "        # Step 4. Selection\n",
    "        # select an agent which has higher fitness \n",
    "        new_population = [candidates[i] if (population[i].fitness >candidates[i].fitness) else population[i] for i in range(0,N)]       \n",
    "        best_agent = sorted(new_population, key=lambda new_population: new_population.fitness)[0]\n",
    "        # update information (not essential part)\n",
    "        val = best_agent.fitness\n",
    "        val_his.append(val)\n",
    "        solution = best_agent.vector\n",
    "        pos_his.append(solution)\n",
    "        \n",
    "        population = new_population\n",
    "        \n",
    "    return solution, val, val_his, pos_his\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### F1\n",
    "![DE_objhistory_f1](report_image/DE_objhistory_f1.png)\n",
    "![DE_cnt_history_f1.png](report_image/DE_cnt_history_f1.png)\n",
    "#### F2\n",
    "![DE_objhistory_f2](report_image/DE_objhistory_f2.png)\n",
    "![DE_cnt_history_f2.png](report_image/DE_cnt_history_f2.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Particle Swarm Optimization (PSO)\n",
    "$v_{i}(k+1) = av_i(k)+bw_1 \\times (p_{b}(k)-x(k))+ cw_2 \\times (p_g(k)-x(k)) $\n",
    "\n",
    "$k$ is itaration number\n",
    "$a$ is inertia weight\n",
    "$b,c$ are learning factors\n",
    "$w$ is radom number\n",
    "\n",
    "![FC_GA](report_image/FC_PSO.png)\n",
    "> cite: Handout Lec09"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "class Particle:\n",
    "    def __init__(self, position, velocity):\n",
    "        self.position = position\n",
    "        self.velocity = velocity\n",
    "        self.fitness = 0.0\n",
    "    \n",
    "    def getPosition(self):\n",
    "        return self.position\n",
    "\n",
    "    def getVelocity(self):\n",
    "        return self.velocity\n",
    "\n",
    "    def setPosition(self, position):\n",
    "        self.position = position\n",
    "\n",
    "    def setVelocity(self,velocity):\n",
    "        self.velocity = velocity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def PSO(obj,dim):\n",
    "\n",
    "    def UpdateParticles(swarm, p_best, g_best):\n",
    "        a = 0.01 # inertia weight\n",
    "        b = 0.01 # the learning factor (personal influence)\n",
    "        c = 0.01 # the learning factor (social finfluence)        \n",
    "\n",
    "        positions = [s.position for s in swarm]    \n",
    "        new_velocities = [a*s.velocity + b*random()*(p_best - s.position) + c*random()*(g_best - s.position)  for s in swarm]\n",
    "        new_positions = [ s.position + v  for (s,v) in zip(swarm, new_velocities)]\n",
    "        new_swarm = [Particle(new_positions[i], new_velocities[i]) for i in range(0,len(swarm))]\n",
    "        return new_swarm\n",
    "\n",
    "    swarm = []  # empty set\n",
    "    N = 50 # The number of swarm\n",
    "    g_best = []\n",
    "    \n",
    "    g_history = []\n",
    "    pos_history = []\n",
    "\n",
    "    # Set the space of finding\n",
    "    if obj == f1:\n",
    "        r = 5.11\n",
    "    elif obj == f2:\n",
    "        r = 2.048\n",
    "    else : \n",
    "        r = 10\n",
    "    \n",
    "    # Step 1: Randomly initialize te swarm\n",
    "    swarm = [Particle(RV4LimitSpace(dim,r) ,random(dim)*10) for i in range(0,N)]\n",
    "\n",
    "    # Step 2: Evaluate all particles\n",
    "    fit_val = [f1(s.position) for s in swarm]\n",
    "    p_best =  swarm[fit_val.index(min(fit_val))].position \n",
    "    g_best = p_best\n",
    "    \n",
    "    cnt = 0\n",
    "    # Step 3: For each particles\n",
    "    while(obj(g_best)>1e-3):\n",
    "        cnt += 1\n",
    "        if cnt > 500 :\n",
    "            print('Could not find')\n",
    "            break\n",
    "            \n",
    "        # Update particles\n",
    "        swarm = UpdateParticles(swarm, p_best, g_best)      \n",
    "        \n",
    "        # Evaluate the particle\n",
    "        fit_val = [f1(s.position) for s in swarm]        \n",
    "        p_best = swarm[fit_val.index(min(fit_val))].position\n",
    "        \n",
    "    # Step 4: Update if necessary the leader of the swarm and the best position obtained by each particles        \n",
    "        if (obj(g_best) < obj(p_best)):\n",
    "            g_best = g_best\n",
    "        else :\n",
    "            g_best = p_best\n",
    "        \n",
    "        g_history.append(obj(g_best))\n",
    "        pos_history.append(g_best)\n",
    "    # Step 5: Stop if terminating condition satisfied; return to Step 3 otherwise          \n",
    "    return g_best, obj(g_best), g_history, pos_history    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### F1\n",
    "![PSO_objhistory_f1](report_image/PSO_objhistory_f1.png)\n",
    "![PSO_cnt_history_f1.png](report_image/PSO_cnt_history_f1.png)\n",
    "#### F2\n",
    "![PSO_objhistory_f2](report_image/PSO_objhistory_f2.png)\n",
    "![PSO_cnt_history_f2.png](report_image/PSO_cnt_history_f2.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## 3. Evaluation\n",
    "\n",
    "### The common condition\n",
    "- the number of trials : 50\n",
    "- criterion: iterate 300 times\n",
    "- population size : 50\n",
    "\n",
    "- #### GA setting\n",
    "    - M = 10  (the lenght of genotype)\n",
    "    - sr = 0.7 (selection rate)\n",
    "    - pm = 0.6 (mutation rate)\n",
    "\n",
    "- #### DE setting\n",
    "    - w = 1 (differential weight [0,2])\n",
    "    - CR = 0.6 (crossover rate [0.4,1])\n",
    "\n",
    "- #### PSO setting\n",
    "    - a = 0.01  (inertia weight)\n",
    "    - b = 0.01  (the learning factor (personal influence) ) \n",
    "    - c = 0.01  (the learning factor (social finfluence)  )\n",
    "\n",
    "#### Accuracy\n",
    "\n",
    "##### F1 \n",
    "|Algorithm|Max.|Min.|Ave.|\n",
    "|:-:|:-:|:-:|:-:|\n",
    "|GA|0.12|0.00|0.03|\n",
    "|DE|0.09|0.00|0.00|\n",
    "|PSO|0.03|0.00|0.02|\n",
    "\n",
    "##### F2\n",
    "|Algorithm|Max.|Min.|Ave.|\n",
    "|:-:|:-:|:-:|:-:|\n",
    "|GA|0.17|0.00|0.02|\n",
    "|DE|0.33|0.00|0.03|\n",
    "|PSO|10.20|0.07|1.73|\n",
    "\n",
    "#### Time [s]\n",
    "##### F1\n",
    "|Algorithm|Max.|Min.|Ave.|\n",
    "|:-:|:-:|:-:|:-:|\n",
    "|GA|2.12|1.5|1.66|\n",
    "|DE|0.40|0.32|0.34|\n",
    "|PSO|0.44|0.22|0.28|\n",
    "\n",
    "##### F2\n",
    "|Algorithm|Max.|Min.|Ave.|\n",
    "|:-:|:-:|:-:|:-:|\n",
    "|GA|2.12|1.13|1.37|\n",
    "|DE|0.11|0.06|0.07|\n",
    "|PSO|0.29|0.15|0.19|\n",
    "\n",
    "#### Memory [MiB]\n",
    "memory_profiler is provided officialy and BSD license.\n",
    "\n",
    "##### F1\n",
    "|Algorithm|Max.|Min.|Ave.|\n",
    "|:-:|:-:|:-:|:-:|\n",
    "|GA|20.12|19.75|19.87|\n",
    "|DE|19.82|19.51|19.64|\n",
    "|PSO|19.83|19.54|19.65|\n",
    "\n",
    "##### F2\n",
    "|Algorithm|Max.|Min.|Ave.|\n",
    "|:-:|:-:|:-:|:-:|\n",
    "|GA|19.98|19.66|19.78|\n",
    "|DE|19.88|19.62|19.76|\n",
    "|PSO|20.09|19.96|19.98|"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "#### Accuracy\n",
    "- GA and DE obtain the global minimum stably.\n",
    "- PSO could not obtain the global minimum sometimes.\n",
    "    - The reason is too few iterations. I checked this algorithm also obtain the global minimum if the iteration number is big enough.\n",
    "\n",
    "#### Time\n",
    "- DE and PSO are the fastest algorithm.\n",
    "- GA has bit-by-bit processing for genotype operation\n",
    "\n",
    "#### Memory\n",
    "There are no big difference about used memory in my implementation.\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
